datafusion/core/src/physical_plan/joins/nested_loop_join.rs

```diff
@@ -0,0 +1,883 @@
+                // just one partition for the left side, and the first partition is all of data for left
+                load_left_specified_partition(0, self.left.clone(), context.clone())
+            })
+        } else {
+            // the distribution of left is not single partition, just need the specified partition for left
+            OnceFut::new(load_left_specified_partition(
+                partition,
+                self.left.clone(),
+                context.clone(),
+            ))
+        };
+        // right side
+        let right_side = if left_is_single_partition {
+            self.right.execute(partition, context)?
+        } else {
+            // the distribution of right is `SinglePartition`
+            self.right.execute(0, context)?
+        };
+
+        Ok(Box::pin(NestedLoopJoinStream {
+            schema: self.schema.clone(),
+            filter: self.filter.clone(),
+            join_type: self.join_type,
+            left_fut,
+            right: right_side,
+            is_exhausted: false,
+            visited_left_side: None,
+            column_indices: self.column_indices.clone(),
+        }))
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => {
+                let display_filter = self.filter.as_ref().map_or_else(
+                    || "".to_string(),
+                    |f| format!(", filter={:?}", f.expression()),
+                );
+                write!(
+                    f,
+                    "NestedLoopJoinExec: join_type={:?}{}",
+                    self.join_type, display_filter
+                )
+            }
+        }
+    }
+
+    fn statistics(&self) -> Statistics {
+        estimate_join_statistics(
+            self.left.clone(),
+            self.right.clone(),
+            vec![],
+            &self.join_type,
+        )
+    }
+}
+
+// For the nested loop join, different join type need the different distribution for
+// left and right node.
+fn distribution_from_join_type(join_type: &JoinType) -> Vec<Distribution> {
+    match join_type {
+        JoinType::Inner | JoinType::Left | JoinType::LeftSemi | JoinType::LeftAnti => {
+            // need the left data, and the right should be one partition
+            vec![
+                Distribution::UnspecifiedDistribution,
+                Distribution::SinglePartition,
+            ]
+        }
+        JoinType::Right | JoinType::RightSemi | JoinType::RightAnti => {
+            // need the right data, and the left should be one partition
+            vec![
+                Distribution::SinglePartition,
+                Distribution::UnspecifiedDistribution,
+            ]
+        }
+        JoinType::Full => {
+            // need the left and right data, and the left and right should be one partition
+            vec![Distribution::SinglePartition, Distribution::SinglePartition]
+        }
+    }
+}
+
```

Comment:
The required distribution logic for the nested loop join is not consistent with the other physical join types(HashJoinExec::CollectLeft, CrossJoinExec), those existing joins do not consider the join types.  For CrossJoinExec and HashJoinExec::CollectLeft, they always collect left. And there is a JoinSelection rule to adjust/reorder the joins sides to make sure the left side is the smaller side.

Reply:
Especially for FullOut join, if we enforce the both input sides coalesced to the single partition,
we might encounter performance issue.

Reply:
In the current execution model, we need to decide the distribution by the `join type`